File: Flutter-Client/flutter-web/lib/data/repositories/download_repository_impl.dart
import 'package:injectable/injectable.dart';
import '../../domain/entities/download.dart';
import '../../domain/repositories/download_repository.dart';
import '../datasources/download_api_client.dart';
import '../datasources/download_websocket_client.dart';
import '../models/download_model.dart';

/// Implementation of DownloadRepository using HTTP API and WebSocket
@LazySingleton(as: DownloadRepository)
class DownloadRepositoryImpl implements DownloadRepository {
  final DownloadApiClient _apiClient;
  final DownloadWebSocketClient _wsClient;

  DownloadRepositoryImpl(this._apiClient, this._wsClient) {
    // Connect to WebSocket on creation
    _wsClient.connect();
  }

  @override
  Future<List<Download>> getActiveDownloads() async {
    try {
      final models = await _apiClient.getQueue();
      return models.map((m) => m.toEntity()).toList();
    } catch (e) {
      throw Exception('Failed to get active downloads: $e');
    }
  }

  @override
  Future<List<Download>> getCompletedDownloads() async {
    try {
      final models = await _apiClient.getCompleted();
      return models.map((m) => m.toEntity()).toList();
    } catch (e) {
      throw Exception('Failed to get completed downloads: $e');
    }
  }

  @override
  Future<List<Download>> getPendingDownloads() async {
    try {
      final models = await _apiClient.getPending();
      return models.map((m) => m.toEntity()).toList();
    } catch (e) {
      throw Exception('Failed to get pending downloads: $e');
    }
  }

  @override
  Future<Download?> getDownloadById(String id) async {
    try {
      // Try to find in all queues
      final queues = await Future.wait([
        getActiveDownloads(),
        getCompletedDownloads(),
        getPendingDownloads(),
      ]);

      for (final queue in queues) {
        final download = queue.where((d) => d.id == id).firstOrNull;
        if (download != null) return download;
      }

      return null;
    } catch (e) {
      throw Exception('Failed to get download by ID: $e');
    }
  }

  @override
  Future<Download> addDownload({
    required String url,
    required String quality,
    required String format,
    String? folder,
    bool autoStart = true,
  }) async {
    try {
      final request = AddDownloadRequest(
        url: url,
        quality: quality,
        format: format,
        folder: folder,
        autoStart: autoStart,
      );

      final response = await _apiClient.addDownload(request);

      if (response.status == 'error' || response.download == null) {
        throw Exception(response.error ?? 'Failed to add download');
      }

      return response.download!.toEntity();
    } catch (e) {
      throw Exception('Failed to add download: $e');
    }
  }

  @override
  Future<void> startDownload(String id) async {
    try {
      await _apiClient.startDownload(id);
    } catch (e) {
      throw Exception('Failed to start download: $e');
    }
  }

  @override
  Future<void> cancelDownload(String id) async {
    try {
      await _apiClient.cancelDownload(id);
    } catch (e) {
      throw Exception('Failed to cancel download: $e');
    }
  }

  @override
  Future<void> deleteDownload(String id) async {
    try {
      await _apiClient.deleteDownload(id);
    } catch (e) {
      throw Exception('Failed to delete download: $e');
    }
  }

  @override
  Future<void> clearCompleted() async {
    try {
      await _apiClient.clearCompleted();
    } catch (e) {
      throw Exception('Failed to clear completed: $e');
    }
  }

  @override
  Stream<Download> watchDownloadUpdates() {
    return _wsClient.updates.map((model) => model.toEntity());
  }

  @override
  Stream<Download> watchDownloadAdditions() {
    return _wsClient.additions.map((model) => model.toEntity());
  }

  @override
  Stream<String> watchDownloadDeletions() {
    return _wsClient.deletions;
  }

  @override
  Stream<void> watchClearedEvents() {
    return _wsClient.cleared;
  }

  /// Check WebSocket connection status
  bool get isConnected => _wsClient.isConnected;

  /// Get WebSocket connection status stream
  Stream<bool> get connectionStatus => _wsClient.connectionStatus;

  /// Dispose resources
  void dispose() {
    _wsClient.dispose();
  }
}

File: Android-Client/app/src/main/java/com/grabtube/android/data/repository/DownloadRepositoryImpl.kt
package com.grabtube.android.data.repository

import com.grabtube.android.data.local.dao.DownloadDao
import com.grabtube.android.data.local.entity.toDomain
import com.grabtube.android.data.local.entity.toEntity
import com.grabtube.android.data.remote.api.GrabTubeApi
import com.grabtube.android.data.remote.dto.AddDownloadRequestDto
import com.grabtube.android.data.remote.dto.toDomain
import com.grabtube.android.data.remote.websocket.SocketManager
import com.grabtube.android.domain.model.Download
import com.grabtube.android.domain.model.DownloadHistory
import com.grabtube.android.domain.model.DownloadRequest
import com.grabtube.android.domain.model.DownloadStatus
import com.grabtube.android.domain.repository.DownloadRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DownloadRepositoryImpl @Inject constructor(
    private val api: GrabTubeApi,
    private val downloadDao: DownloadDao,
    private val socketManager: SocketManager
) : DownloadRepository {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    init {
        // Observe socket events and update local database
        scope.launch {
            socketManager.observeDownloadEvents().collect { event ->
                when (event) {
                    is SocketManager.DownloadEvent.Added -> {
                        val download = event.download.toDomain()
                        downloadDao.insert(download.toEntity())
                    }
                    is SocketManager.DownloadEvent.Updated -> {
                        val download = event.download.toDomain()
                        downloadDao.insert(download.toEntity())
                    }
                    is SocketManager.DownloadEvent.Completed -> {
                        downloadDao.getById(event.downloadId)?.let { entity ->
                            val updated = entity.copy(
                                status = DownloadStatus.COMPLETED.name,
                                completedAt = System.currentTimeMillis() / 1000
                            )
                            downloadDao.update(updated)
                        }
                    }
                    is SocketManager.DownloadEvent.Canceled -> {
                        downloadDao.getById(event.downloadId)?.let { entity ->
                            val updated = entity.copy(status = DownloadStatus.CANCELED.name)
                            downloadDao.update(updated)
                        }
                    }
                    is SocketManager.DownloadEvent.Cleared -> {
                        downloadDao.deleteAll()
                    }
                }
            }
        }
    }

    override fun observeDownloads(): Flow<List<Download>> {
        return downloadDao.observeAll().map { entities ->
            entities.map { it.toDomain() }
        }
    }

    override fun observeDownload(id: String): Flow<Download?> {
        return downloadDao.observeById(id).map { it?.toDomain() }
    }

    override fun getHistory(limit: Int, offset: Int): Flow<List<DownloadHistory>> {
        return downloadDao.getHistory(limit, offset).map { entities ->
            entities.map { entity ->
                val download = entity.toDomain()
                DownloadHistory(
                    id = download.id,
                    url = download.url,
                    title = download.title,
                    thumbnail = download.thumbnail,
                    quality = download.quality,
                    format = download.format,
                    folder = download.folder,
                    status = download.status,
                    completedAt = download.completedAt ?: download.addedAt,
                    duration = download.completedAt?.let { completed ->
                        download.startedAt?.let { started ->
                            completed.epochSecond - started.epochSecond
                        }
                    },
                    fileSize = download.totalBytes
                )
            }
        }
    }

    override suspend fun addDownload(request: DownloadRequest): Result<Download> {
        return try {
            val requestDto = AddDownloadRequestDto(
                url = request.url,
                quality = request.quality,
                format = request.format,
                folder = request.folder,
                autoStart = request.autoStart
            )

            val response = api.addDownload(requestDto)

            if (response.status == "ok" && response.id != null) {
                // Fetch the full download details
                syncWithServer()
                val download = downloadDao.getById(response.id)?.toDomain()
                    ?: return Result.failure(Exception("Download not found after adding"))
                Result.success(download)
            } else {
                Result.failure(Exception(response.error ?: "Failed to add download"))
            }
        } catch (e: Exception) {
            Timber.e(e, "Failed to add download")
            Result.failure(e)
        }
    }

    override suspend fun startDownload(id: String): Result<Unit> {
        return try {
            api.startDownload(id)
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to start download")
            Result.failure(e)
        }
    }

    override suspend fun cancelDownload(id: String): Result<Unit> {
        return try {
            api.cancelDownload(id)
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to cancel download")
            Result.failure(e)
        }
    }

    override suspend fun pauseDownload(id: String): Result<Unit> {
        // Note: Pause functionality may not be supported by the backend
        // This is a placeholder for future implementation
        return Result.failure(UnsupportedOperationException("Pause not yet supported"))
    }

    override suspend fun resumeDownload(id: String): Result<Unit> {
        // Resume is equivalent to start for now
        return startDownload(id)
    }

    override suspend fun deleteDownload(id: String): Result<Unit> {
        return try {
            api.deleteDownload(id)
            downloadDao.deleteById(id)
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to delete download")
            Result.failure(e)
        }
    }

    override suspend fun clearCompleted(): Result<Unit> {
        return try {
            api.clearCompleted()
            downloadDao.deleteCompleted()
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to clear completed")
            Result.failure(e)
        }
    }

    override suspend fun clearAll(): Result<Unit> {
        return try {
            api.clearAll()
            downloadDao.deleteAll()
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to clear all")
            Result.failure(e)
        }
    }

    override suspend fun retryDownload(id: String): Result<Unit> {
        return try {
            // Delete the failed download and re-add it
            val download = downloadDao.getById(id)?.toDomain()
                ?: return Result.failure(Exception("Download not found"))

            deleteDownload(id)
            addDownload(
                DownloadRequest(
                    url = download.url,
                    quality = download.quality,
                    format = download.format,
                    folder = download.folder,
                    autoStart = true
                )
            )
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to retry download")
            Result.failure(e)
        }
    }

    override suspend fun syncWithServer(): Result<Unit> {
        return try {
            val downloads = api.getDownloads()
            val entities = downloads.map { it.toDomain().toEntity() }
            downloadDao.insertAll(entities)
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Failed to sync with server")
            Result.failure(e)
        }
    }
}
